### PR TITLE
[#1297] Upgrade to Spring Boot 2

### PR DESCRIPTION
**Pull request by kaniyan**
**Not to merge until necessary CQs are created and approved.**

1. The spring artifacts are upgraded as below.

| **Artifact** | **Version** |
| ------------- | ------------- |
| Spring Boot | 2.1.5.RELEASE |
| Spring  | 5.1.8.RELEASE |
| Spring  Security | 5.1.5.RELEASE |

2. The below dependencies/plug-ins are updated to be in compatible with the above spring version.

| **Artifact** | **Version** |
| ------------- | ------------- |
| snakeyaml | 1.23 |
| maven-deploy-plugin  | 2.8.2 |
| maven-assembly-plugin |3.1.0 |

3. Support for Guava has been [dropped](https://github.com/spring-projects/spring-boot/wiki/Spring-Boot-2.0-Migration-Guide#guavacachemanager) in Spring Framework 5. Instead of `GuavaCacheManager`,  `CaffeineCacheManager` is being used and the corresponding dependency ([com.github.ben-manes.caffeine:caffeine:2.7.0](https://github.com/ben-manes/caffeine/tree/v2.7.0)) is added. It uses Apache 2.0 license.

4. [Changed](https://docs.spring.io/spring-boot/docs/2.1.5.RELEASE/reference/html/boot-features-external-config.html#boot-features-external-config-relaxed-binding) the prefix value in annotations to be in kebab case (lowercase and separated by -)

5. Spring Boot Maven plug-in configuration attributes are to [start](https://github.com/spring-projects/spring-boot/wiki/Spring-Boot-2.0-Migration-Guide#spring-boot-maven-plugin) with `spring-boot` as in the below example. It has been updated in the documentation.
```
mvn spring-boot:run -Dspring-boot.run.arguments=--hono.client.username=consumer@HONO,--hono.client.password=verysecret -Dspring-boot.run.profiles=command,ssl
```

### CODE REVIEW

File: bom/pom.xml [hunk 1]
```diff
@@ -34,6 +34,7 @@
     <dispatch-router.image.name>quay.io/enmasse/qdrouterd-base:1.7.0</dispatch-router.image.name>
     <grafana.version>5.3.2</grafana.version>
     <guava.version>25.0-jre</guava.version>
+    <caffeine.version>2.7.0</caffeine.version>
```
Comment (sophokles73) [inline bom/pom.xml]: the idea is to have the list in alphabetical order ...
BTW does it also work with version 2.6.2? That would be easier to get approved ...

File: legal/legal/LICENSE.Spring.txt [hunk 2]
```diff
@@ -202,9 +202,9 @@
 
 =======================================================================
 
-SPRING FRAMEWORK 4.3.18.RELEASE SUBCOMPONENTS:
+SPRING FRAMEWORK 5.1.8.RELEASE SUBCOMPONENTS:
```
Comment (sophokles73) [inline legal/legal/LICENSE.Spring.txt]: did you just change the version in the existing file or did you replace the existing file with the corresponding file from the 5.1.8 artifacts?

File: legal/legal/NOTICE.Spring.txt [hunk 3]
```diff
@@ -1,5 +1,5 @@
-Spring Framework 4.3.18.RELEASE
-Copyright (c) 2002-2018 Pivotal, Inc.
+Spring Framework 5.1.8.RELEASE
+Copyright (c) 2002-2019 Pivotal, Inc.
```
Comment (sophokles73) [inline legal/legal/NOTICE.Spring.txt]: did you just change the version in the existing file or did you replace the existing file with the corresponding file from the 5.1.8 artifacts?

File: service-base/src/main/java/org/eclipse/hono/service/AbstractAdapterConfig.java [hunk 4]
```diff
@@ -455,8 +455,8 @@ public HealthCheckServer healthCheckServer() {
      * @return The properties.
      */
     @Bean
-    @ConfigurationProperties(prefix = "hono.plan.prometheusBased")
-    @ConditionalOnProperty(name = "hono.plan.prometheusBased.host")
+    @ConfigurationProperties(prefix = "hono.plan.prometheus-based")
```
Comment (sophokles73) [inline service-base/src/main/java/org/eclipse/hono/service/AbstractAdapterConfig.java]: these properties are defined in the documentation, which you'll need to adapt accordingly as well ...

Comment (kaniyan) [reply service-base/src/main/java/org/eclipse/hono/service/AbstractAdapterConfig.java]: The above changes are needed only in the annotation side. The property `hono.plan.prometheus-based.host` or  `hono.plan.prometheusBased.host` can all be used as Spring Boot supports relaxed binding. So the properties such as  `hono.plan.prometheusBased.host` are still valid.

Comment (sophokles73) [reply service-base/src/main/java/org/eclipse/hono/service/AbstractAdapterConfig.java]: That might be true. Still, our documentation should define the property name that we are actually using in the code, don't you think? It is an added bonus that people who (still) use a similar but different property also succeed ...

Comment (kaniyan) [reply service-base/src/main/java/org/eclipse/hono/service/AbstractAdapterConfig.java]: yes, I will update it in the documentation too.

File: legal/legal/NOTICE.Spring.txt (hunk 3 above)
Comment (kaniyan) [reply legal/legal/NOTICE.Spring.txt]: I followed the same as for the spring 4.3.12.RELEASE update. I changed the version and the year in the above file. Also I compared it with the _notice.txt_ from the `spring-core-5.1.8.RELEASE.jar` artifact.

Comment (sophokles73) [reply legal/legal/NOTICE.Spring.txt]: My point is, we cannot just update the year but indeed need to replace the existing file with the one from the 5.1.8.RELEASE jar (if it differs in anything else than just the year).

Comment (kaniyan) [reply legal/legal/NOTICE.Spring.txt]: There is no difference from the file in 5.1.8.RELEASE jar except the version and the date.

File: bom/pom.xml (hunk 1 above)
Comment (kaniyan) [reply bom/pom.xml]: It works with version 2.6.2. I will push the changes.

File: legal/legal/LICENSE.Spring.txt (hunk 2 above)
Comment (kaniyan) [reply legal/legal/LICENSE.Spring.txt]: I have modified the existing file. In case of license file, I will replace it with the file from 5.1.8 artifact.

**Comment by sophokles73** [comment]
@kaniyan can you rebase to resolve conflicts?

**Comment by kaniyan** [comment]
> @kaniyan can you rebase to resolve conflicts?

@sophokles73 rebased.

File: service-base/src/main/java/org/eclipse/hono/service/AbstractAdapterConfig.java (hunk 4 above)
Comment (sophokles73) [reply service-base/src/main/java/org/eclipse/hono/service/AbstractAdapterConfig.java]: I wasn't aware that this is required by Spring 2. I thought you deliberately chose to change the property prefix. Based on my new knowledge I think that we should keep the original camel case notation in our documentation for consistency. Would you mind changing back the property names in the docs to the original values? Sorry for the inconvenience ...

Comment (kaniyan) [reply service-base/src/main/java/org/eclipse/hono/service/AbstractAdapterConfig.java]: I have removed the changes regarding property names in the docs.

**Comment by sophokles73** [comment]
Great work, @kaniyan 
Thanks a lot!

**Pull request merged**